src/main/java/org/junitpioneer/jupiter/report/PioneerReportListener.java

```diff
@@ -0,0 +1,199 @@
+import org.junit.platform.launcher.TestIdentifier;
+import org.junit.platform.launcher.TestPlan;
+
+/**
+ * <p>This listener creates an additional report for pioneer extension and validates it against the "pioneerreport.xsd".
+ * This is done because the default junit test report can't / should not be extended.
+ * A dynamic test report which shall be used by all test tools is planed, but not yet realized, see
+ * <a href="https://github.com/ota4j-team/opentest4j/issues/9">issue #9 at opentest4j</a> for current progress.</p>
+ *
+ * <p>The report contains test containers and its test methods. For each object properties, based on published
+ * report entries are reported. This allows extension to report additional values in the report.
+ * for test cases the result of the test is (of course) reported.</p>
+ *
+ * @see <a href="https://junit.org/junit5/docs/current/api/org.junit.platform.launcher/org/junit/platform/launcher/TestPlan.html">JUnit Docs</a>
+ */
+public class PioneerReportListener implements TestExecutionListener {
+
+	TestPlan storedTestPlan;
+	PioneerReport pioneerReport;
+
+	ConcurrentHashMap<String, List<Property>> propertyCache = new ConcurrentHashMap<>();
+	ConcurrentHashMap<String, String> testStatusCache = new ConcurrentHashMap<>();
+
+	@Override
+	public void reportingEntryPublished(TestIdentifier testIdentifier, ReportEntry entry) {
+		String id = testIdentifier.getUniqueId();
+
+		// Store published entries in cache
+		propertyCache.putIfAbsent(id, new ArrayList<>());
+		entry.getKeyValuePairs().forEach((key, value) -> propertyCache.get(id).add(new Property(key, value)));
+	}
+
+	@Override
+	public void executionFinished(TestIdentifier testIdentifier, TestExecutionResult testExecutionResult) {
+		if (testIdentifier.isTest()) {
+			// Store result in cache
+			testStatusCache.put(testIdentifier.getUniqueId(), testExecutionResult.getStatus().toString());
+		}
+	}
+
+	@Override
+	public void testPlanExecutionFinished(TestPlan testPlan) {
+		// Store plan to access it from methods
+		this.storedTestPlan = testPlan;
+		pioneerReport = new PioneerReport();
+
+		Set<TestIdentifier> allRoots = testPlan.getRoots();
+
+		for (TestIdentifier identifier : allRoots) {
+			if (identifier.isContainer()) {
+				pioneerReport.getTestContainer().add(processContainer(identifier));
+			} else if (identifier.isTest()) {
+				pioneerReport.getTestCase().add(processTest(identifier));
+			}
+		}
+		writeReport(pioneerReport);
+	}
+
+	/**
+	 * Converts an {@code TestIdentifier} object into a {@code Testcontainer} for the report.
+	 *
+	 * @param container Object to be converted into {@code Testcontainer}
+	 * @return Created object
+	 */
+	TestContainer processContainer(TestIdentifier container) {
+		TestContainer testcontainer = new TestContainer(container.getDisplayName());
+
+		// Add properties, if there are any
+		String uniqueId = container.getUniqueId();
+		Properties allProps = getProperties(uniqueId);
```

Comment:
If `getProperties` returned with `Optional<Properties>`, this whole thing could look like this:
```
String uniqueId = container.getUniqueId();
// Add properties, if there are any present
getProperties(uniqueId).ifPresent(testContainer::setProperties);
```